File: py/aoc2022/day17.py
"""
Day 17: Pyroclastic Flow
"""

ROCKS = (
    (4, (15,)),
    (3, (2, 7, 2)),
    (3, (7, 4, 4)),
    (1, (1, 1, 1, 1)),
    (2, (3, 3)),
)


def _check_free(rock, lines, x, y):
    return all(not (line & row << x) for line, row in zip(lines[y:], rock))


def _step(width, rock, jet, jet_index, height, lines):
    # pylint: disable=too-many-arguments
    x, y = 2, len(lines) + 3
    while _check_free(rock, lines, x, y):
        match jet[jet_index]:
            case "<":
                x2 = x - 1
            case ">":
                x2 = x + 1
            case _:
                raise ValueError()
        jet_index = (jet_index + 1) % len(jet)
        if 0 <= x2 <= 7 - width and _check_free(rock, lines, x2, y):
            x = x2
        y -= 1
    y, lines = y + 1, list(lines)
    for rock_y, row in enumerate(rock):
        if y + rock_y >= len(lines):
            lines.append(row << x)
            height += 1
        else:
            lines[y + rock_y] = lines[y + rock_y] | row << x
    visible, queue = [0] * len(lines) + [1], [(0, len(lines))]
    while queue:
        x, y = queue.pop()
        for x2, y2 in ((x, y - 1), (x - 1, y), (x + 1, y), (x, y + 1)):
            if (
                x2 in range(7)
                and y2 in range(len(visible))
                and not (visible[y2] & 1 << x2)
            ):
                visible[y2] |= 1 << x2
                if not (y2 < len(lines) and lines[y2] & 1 << x2):
                    queue.append((x2, y2))
    lines = [line & row for line, row in zip(lines, visible)]
    trim = 0
    while not lines[trim]:
        trim += 1
    return jet_index, height, lines[trim:]


def _solve(jet, count):
    jet_index, height, lines = 0, 0, [127]
    seen, heights = {}, []
    for i in range(count):
        rock_index = i % len(ROCKS)
        width, rock = ROCKS[rock_index]
        j = seen.setdefault((tuple(lines), jet_index, rock_index), i)
        if j < i:
            return heights[j + (count - j) % (i - j)] + (count - j) // (i - j) * (
                height - heights[j]
            )
        heights.append(height)
        jet_index, height, lines = _step(width, rock, jet, jet_index, height, lines)
    return height


def part1(lines):
    """
    >>> part1([">>><<><>><<<>><>>><<<>>><<<><<<>><>><<>>"])
    3068
    """
    return _solve(next(iter(lines)).strip(), 2022)


def part2(lines):
    """
    >>> part2([">>><<><>><<<>><>>><<<>>><<<><<<>><>><<>>"])
    1514285714288
    """
    return _solve(next(iter(lines)).strip(), 1000000000000)


parts = (part1, part2)

File: rs/src/day17.rs
use smallvec::{smallvec, SmallVec};
use static_init::dynamic;
use std::collections::HashMap;

fn parse<'a, I, S>(lines: I) -> Vec<i8>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    lines
        .into_iter()
        .flat_map(|line| line.as_ref().chars())
        .filter_map(|char| match char {
            '<' => Some(-1),
            '>' => Some(1),
            _ => None,
        })
        .collect()
}

struct Rock {
    width: u8,
    rows: SmallVec<[u8; 4]>,
}
#[dynamic]
static ROCKS: [Rock; 5] = [
    Rock {
        width: 4,
        rows: smallvec![15],
    },
    Rock {
        width: 3,
        rows: smallvec![2, 7, 2],
    },
    Rock {
        width: 3,
        rows: smallvec![7, 4, 4],
    },
    Rock {
        width: 1,
        rows: smallvec![1, 1, 1, 1],
    },
    Rock {
        width: 2,
        rows: smallvec![3, 3],
    },
];

struct State {
    height: u64,
    visible: Vec<u8>,
}
impl State {
    fn contains(&self, x: u8, y: usize, rock: &Rock) -> bool {
        rock.rows
            .iter()
            .enumerate()
            .filter_map(|(dy, row)| Some(self.visible.get(y + dy)? & row << x))
            .any(|intersection| intersection != 0)
    }

    fn add(&self, x: u8, y: usize, rock: &Rock) -> State {
        assert!(y <= self.visible.len());
        let mut height = self.height;
        let mut visible = self.visible.clone();
        for (dy, row) in rock.rows.iter().enumerate() {
            if let Some(v) = visible.get_mut(y + dy) {
                *v |= row << x;
            } else {
                height += 1;
                visible.push(row << x);
            }
        }

        let mut seen = vec![0; visible.len() + 1];
        *seen.last_mut().unwrap() = 1;
        let mut queue = vec![visible.len() << 3];
        while let Some(pos) = queue.pop() {
            if pos < visible.len() << 3 && visible[pos >> 3] & 1 << (pos & 7) != 0 {
                continue;
            }
            for pos in [
                pos.checked_sub(8),
                pos.checked_sub(1).filter(|&pos| pos & 7 != 7),
                Some(pos + 1).filter(|&pos| pos & 7 != 7),
                Some(pos + 8).filter(|&pos| pos < seen.len() << 3),
            ] {
                let Some(pos) = pos else { continue };
                if seen[pos >> 3] & 1 << (pos & 7) == 0 {
                    seen[pos >> 3] |= 1 << (pos & 7);
                    queue.push(pos);
                }
            }
        }

        for (v, s) in visible.iter_mut().zip(seen.iter()) {
            *v &= s;
        }
        visible.drain(..visible.iter().enumerate().find(|(_, &v)| v != 0).unwrap().0);
        State { height, visible }
    }
}

fn solve(jet: &[i8], n: u64) -> u64 {
    let mut jet_index = 0;
    let mut state = State {
        height: 0,
        visible: vec![127],
    };
    let mut seen = HashMap::new();
    let mut heights = vec![];
    for i in 0..n {
        let rock_index = (i % ROCKS.len() as u64) as usize;
        if let Some(j) = seen.insert((rock_index, jet_index, state.visible.clone()), i) {
            let q = (n - j) / (i - j);
            let r = (n - j) % (i - j);
            return heights[(j + r) as usize] + q * (state.height - heights[j as usize]);
        }
        let rock = &ROCKS[rock_index];
        let (mut x, mut y) = (2u8, state.visible.len() + 3);
        while !state.contains(x, y, rock) {
            x = x
                .checked_add_signed(jet[jet_index])
                .filter(|&x| x + rock.width <= 7 && !state.contains(x, y, rock))
                .unwrap_or(x);
            jet_index = (jet_index + 1) % jet.len();
            y -= 1;
        }
        heights.push(state.height);
        state = state.add(x, y + 1, rock);
    }
    state.height
}

pub fn part1<'a, I, S>(lines: I) -> u64
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    solve(&parse(lines), 2022)
}

pub fn part2<'a, I, S>(lines: I) -> u64
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    solve(&parse(lines), 1000000000000)
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &[&str] = &[">>><<><>><<<>><>>><<<>>><<<><<<>><>><<>>"];

    #[test]
    fn part1_examples() {
        assert_eq!(3068, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(1514285714288, part2(EXAMPLE));
    }
}
